### DISCUSSION TITLE
BOM, Substitutes, Build Order Report and context variables

### DISCUSSION THREAD

**Discussion by SergeoLacruz** [inventree/InvenTree#2847] [Q&A] [answered]
Hi,
After importing of 2500 components from an old library I now started to work on the output side.
The substitutes function in BOM is a very useful feature for last minute changes or in case of
component shortage. But I have a problem with exporting those. 
In a BOM export the substitutes do not show up. I think this is OK because they might belong to
the build order. There I do final definition using the stock allocation. 
Using the build order report template dour not report substitutes. So I started to play with the
context variables, documented and undocumented ones:

[ https://github.com/SergeoLacruz/inventree-docs/blob/master/docs/report/build.md](
https://github.com/SergeoLacruz/inventree-docs/blob/master/docs/report/build.md)

I could not dig any deeper so far. Is it possible to document the substitutes in that way or
is there another (better) solution? Thanks in advance,

Michael

**Comment by matmair** [comment]
@SergeoLacruz the reporting feature is somewhat undocumented — it might be that the substitutes are not exposed. That is not a problem, we can easily add that, should a 1-liner

**Comment by SchrodingersGat** [comment, marked as answer]
@SergeoLacruz I would very much appreciate a PR against the documentation repo based on what you have there - it looks great!

Regarding substitutes, you *should* be able to do something like:

```
{% for line in build.bom_items %}
{% for sub in line.substitutes %}
...
{% endfor %}
{% endfor %}
```

**Comment by SergeoLacruz** [comment]
Thanks for the hint. I will try this later the day.

**Comment by SergeoLacruz** [comment]
line.substitutes.all does the trick. I created a first example. 

https://github.com/SergeoLacruz/inventree-docs/blob/master/docs/assets/images/report/bom_example.png

To me this looks like a very powerfull system for documentation. I think we can more or less automatically create what we call a build instruction. I will describe this BOM example soon.
